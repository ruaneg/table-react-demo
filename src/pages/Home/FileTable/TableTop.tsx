import styled from 'styled-components';
import DownloadButton from '~/components/Buttons/DownloadButton';
import { FileData } from './fileData';

export const StyledTableTop = styled.div`
  padding: 16px;
  border: 1px solid #f1efef;
  border-bottom: none;
  font-size: 1rem;
  display: flex;
`;

const SelectedText = styled.div`
  width: 8rem;
  margin: 0 1rem;
`;

export const TableTop = ({
  selectAll,
  indeterminate,
  selectedRows,
  handleSelectAllChange,
}: {
  selectAll: boolean;
  indeterminate: boolean;
  selectedRows: FileData[];
  handleSelectAllChange: () => void;
}) => {
  const handleDownloadSelected = () => {
    const selectedPaths = selectedRows
      .map((file) => `${file.device}: ${file.path}`)
      .join('\n');
    alert(`Selected files:\n${selectedPaths}`);
  };

  return (
    <StyledTableTop>
      <input
        aria-label="Select all available files"
        type="checkbox"
        checked={selectAll}
        ref={(input) => {
          if (input) input.indeterminate = indeterminate;
        }}
        onChange={handleSelectAllChange}
        style={{ marginLeft: '-4px' }}
      />
      <SelectedText>
        {selectedRows.length > 0
          ? `Selected ${selectedRows.length}`
          : 'None Selected'}
      </SelectedText>
      <div>
        <DownloadButton
          aria-label={'Download Selected files'}
          label={'Download Selected'}
          handleOnClick={handleDownloadSelected}
          disabled={selectedRows.length === 0}
        />
      </div>
    </StyledTableTop>
  );
};
