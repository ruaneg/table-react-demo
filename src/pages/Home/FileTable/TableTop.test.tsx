import { render, screen, fireEvent } from "@testing-library/react";
import "jest-styled-components";
import { describe, it, expect, vi } from "vitest";
import { TableTop } from "./TableTop";
import { FileData } from "./fileData";

describe("TableTop", () => {
  const mockFiles: FileData[] = [
    {
      name: "smss.exe",
      device: "Mario",
      path: "\\Device\\HarddiskVolume2\\Windows\\System32\\smss.exe",
      status: "scheduled",
    },
    {
      name: "netsh.exe",
      device: "Luigi",
      path: "\\Device\\HarddiskVolume2\\Windows\\System32\\netsh.exe",
      status: "available",
    },
  ];

  const handleSelectAllChange = vi.fn();

  test("renders correctly with initial props", () => {
    const { container } = render(
      <TableTop
        selectAll={false}
        indeterminate={false}
        selectedRows={[]}
        handleSelectAllChange={handleSelectAllChange}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('displays "None Selected" when no files are selected', () => {
    render(
      <TableTop
        selectAll={false}
        indeterminate={false}
        selectedRows={[]}
        handleSelectAllChange={handleSelectAllChange}
      />
    );
    expect(screen.getByText("None Selected")).toBeInTheDocument();
  });

  test("displays the correct number of selected files", () => {
    render(
      <TableTop
        selectAll={false}
        indeterminate={false}
        selectedRows={mockFiles}
        handleSelectAllChange={handleSelectAllChange}
      />
    );
    expect(
      screen.getByText(`Selected ${mockFiles.length}`)
    ).toBeInTheDocument();
  });

  test("calls handleSelectAllChange when the checkbox is clicked", () => {
    render(
      <TableTop
        selectAll={false}
        indeterminate={false}
        selectedRows={mockFiles}
        handleSelectAllChange={handleSelectAllChange}
      />
    );
    const checkbox = screen.getByLabelText("Select all available files");
    fireEvent.click(checkbox);
    expect(handleSelectAllChange).toHaveBeenCalledTimes(1);
  });

  test("disables the download button when no files are selected", () => {
    render(
      <TableTop
        selectAll={false}
        indeterminate={false}
        selectedRows={[]}
        handleSelectAllChange={handleSelectAllChange}
      />
    );
    const button = screen.getByLabelText("Download Selected files");
    expect(button).toBeDisabled();
  });

  test("enables the download button when files are selected", () => {
    render(
      <TableTop
        selectAll={false}
        indeterminate={false}
        selectedRows={mockFiles}
        handleSelectAllChange={handleSelectAllChange}
      />
    );
    const button = screen.getByLabelText("Download Selected files");
    expect(button).not.toBeDisabled();
  });
});
