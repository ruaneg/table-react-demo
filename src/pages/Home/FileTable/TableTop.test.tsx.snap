// Vitest Snapshot v1

exports[`TableTop > renders correctly with initial props 1`] = `
.c3 {
  padding: 0 8px;
  font-size: 0.875rem;
}

.c2 {
  border: 0;
  background: none;
  box-shadow: none;
  border-radius: 0px;
  font-size: 1rem;
  color: #000000d6;
  cursor: pointer;
}

.c2:active {
  color: dodgerblue;
}

.c2:disabled {
  cursor: not-allowed;
  opacity: 0.6;
}

.c0 {
  padding: 16px;
  border: 1px solid #f1efef;
  border-bottom: none;
  font-size: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  width: 8rem;
  margin: 0 1rem;
}

<div
  class="c0"
>
  <input
    aria-label="Select all available files"
    style="margin-left: -4px;"
    type="checkbox"
  />
  <div
    class="c1"
  >
    None Selected
  </div>
  <div>
    <button
      aria-label="Download Selected files"
      class="c2"
      disabled=""
    >
      <span
        class="c3"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M216 0h80c13.3 0 24 10.7 24 24v168h87.7c17.8 0 26.7 21.5 14.1 34.1L269.7 378.3c-7.5 7.5-19.8 7.5-27.3 0L90.1 226.1c-12.6-12.6-3.7-34.1 14.1-34.1H192V24c0-13.3 10.7-24 24-24zm296 376v112c0 13.3-10.7 24-24 24H24c-13.3 0-24-10.7-24-24V376c0-13.3 10.7-24 24-24h146.7l49 49c20.1 20.1 52.5 20.1 72.6 0l49-49H488c13.3 0 24 10.7 24 24zm-124 88c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20zm64 0c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20z"
          />
        </svg>
      </span>
      Download Selected
    </button>
  </div>
</div>
`;
